File: src/detectors/macleod.js

module.exports = function(config) {

  config = config || {};

  /**
   * The expected size of an audio buffer (in samples).
   */
  const DEFAULT_BUFFER_SIZE = 1024;

  /**
   * Defines the relative size the chosen peak (pitch) has. 0.93 means: choose
   * the first peak that is higher than 93% of the highest peak detected. 93%
   * is the default value used in the Tartini user interface.
   */
  const DEFAULT_CUTOFF = 0.97;

  const DEFAULT_SAMPLE_RATE = 44100;

  /**
   * For performance reasons, peaks below this cutoff are not even considered.
   */
  const SMALL_CUTOFF = 0.5;

  /**
   * Pitch annotations below this threshold are considered invalid, they are
   * ignored.
   */
  const LOWER_PITCH_CUTOFF = 80;

  /**
   * Defines the relative size the chosen peak (pitch) has.
   */
  const cutoff = config.cutoff || DEFAULT_CUTOFF;

  /**
   * The audio sample rate. Most audio has a sample rate of 44.1kHz.
   */
  const sampleRate = config.sampleRate || DEFAULT_SAMPLE_RATE;

  /**
   * Size of the input buffer.
   */
  const bufferSize = config.bufferSize || DEFAULT_BUFFER_SIZE;

  /**
   * Contains a normalized square difference function value for each delay
   * (tau).
   */
  const nsdf = new Float32Array(bufferSize);

  /**
   * The x and y coordinate of the top of the curve (nsdf).
   */
  let turningPointX;
  let turningPointY;

  /**
   * A list with minimum and maximum values of the nsdf curve.
   */
  let maxPositions = [];

  /**
   * A list of estimates of the period of the signal (in samples).
   */
  let periodEstimates = [];

  /**
   * A list of estimates of the amplitudes corresponding with the period
   * estimates.
   */
  let ampEstimates = [];

  /**
   * The result of the pitch detection iteration.
   */
  const result = {};

  /**
   * Implements the normalized square difference function. See section 4 (and
   * the explanation before) in the MPM article. This calculation can be
   * optimized by using an FFT. The results should remain the same.
   */
  const normalizedSquareDifference = function(float32AudioBuffer) {
    for (let tau = 0; tau < float32AudioBuffer.length; tau++) {
      let acf = 0;
      let divisorM = 0;
      for (let i = 0; i < float32AudioBuffer.length - tau; i++) {
        acf += float32AudioBuffer[i] * float32AudioBuffer[i+tau];
        divisorM += float32AudioBuffer[i] * float32AudioBuffer[i] + float32AudioBuffer[i + tau] * float32AudioBuffer[i + tau];
      }
      nsdf[tau] = 2 * acf / divisorM;
    }
  };

  /**
   * Finds the x value corresponding with the peak of a parabola.
   * Interpolates between three consecutive points centered on tau.
   */
  const parabolicInterpolation = function(tau) {
    const nsdfa = nsdf[tau - 1],
        nsdfb = nsdf[tau],
        nsdfc = nsdf[tau + 1],
        bValue = tau,
        bottom = nsdfc + nsdfa - 2 * nsdfb;
    if (bottom === 0) {
      turningPointX = bValue;
      turningPointY = nsdfb;
    } else {
      const delta = nsdfa - nsdfc;
      turningPointX = bValue + delta / (2 * bottom);
      turningPointY = nsdfb - delta * delta / (8 * bottom);
    }
  };

  // Finds the highest value between each pair of positive zero crossings.
  const peakPicking = function() {
    let pos = 0;
    let curMaxPos = 0;

    // find the first negative zero crossing.
    while (pos < (nsdf.length - 1) / 3 && nsdf[pos] > 0) {
      pos++;
    }

    // loop over all the values below zero.
    while (pos < nsdf.length - 1 && nsdf[pos] <= 0) {
      pos++;
    }

    // can happen if output[0] is NAN
    if (pos == 0) {
      pos = 1;
    }

    while (pos < nsdf.length - 1) {
      if (nsdf[pos] > nsdf[pos - 1] && nsdf[pos] >= nsdf[pos + 1]) {
        if (curMaxPos == 0) {
          // the first max (between zero crossings)
          curMaxPos = pos;
        } else if (nsdf[pos] > nsdf[curMaxPos]) {
          // a higher max (between the zero crossings)
          curMaxPos = pos;
        }
      }
      pos++;
      // a negative zero crossing
      if (pos < nsdf.length - 1 && nsdf[pos] <= 0) {
        // if there was a maximum add it to the list of maxima
        if (curMaxPos > 0) {
          maxPositions.push(curMaxPos);
          curMaxPos = 0; // clear the maximum position, so we start
          // looking for a new ones
        }
        while (pos < nsdf.length - 1 && nsdf[pos] <= 0) {
          pos++; // loop over all the values below zero
        }
      }
    }
    if (curMaxPos > 0) {
      maxPositions.push(curMaxPos);
    }
  };

  return function(float32AudioBuffer) {

    // 0. Clear old results.
    let pitch;
    maxPositions = [];
    periodEstimates = [];
    ampEstimates = [];

    // 1. Calculute the normalized square difference for each Tau value.
    normalizedSquareDifference(float32AudioBuffer);
    // 2. Peak picking time: time to pick some peaks.
    peakPicking();

    let highestAmplitude = -Infinity;

    for (let i = 0; i < maxPositions.length; i++) {
      const tau = maxPositions[i];
      // make sure every annotation has a probability attached
      highestAmplitude = Math.max(highestAmplitude, nsdf[tau]);

      if (nsdf[tau] > SMALL_CUTOFF) {
        // calculates turningPointX and Y
        parabolicInterpolation(tau);
        // store the turning points
        ampEstimates.push(turningPointY);
        periodEstimates.push(turningPointX);
        // remember the highest amplitude
        highestAmplitude = Math.max(highestAmplitude, turningPointY);
      }
    }

    if (periodEstimates.length) {
      // use the overall maximum to calculate a cutoff.
      // The cutoff value is based on the highest value and a relative
      // threshold.
      const actualCutoff = cutoff * highestAmplitude;
      let periodIndex = 0;

      for (let i = 0; i < ampEstimates.length; i++) {
        if (ampEstimates[i] >= actualCutoff) {
          periodIndex = i;
          break;
        }
      }

      const period = periodEstimates[periodIndex],
          pitchEstimate = sampleRate / period;

      if (pitchEstimate > LOWER_PITCH_CUTOFF) {
        pitch = pitchEstimate;
      } else {
        pitch = -1;
      }

    } else {
      // no pitch detected.
      pitch = -1;
    }

    result.probability = highestAmplitude;
    result.freq = pitch;
    return result;
  };
};

File: src/detectors/macleod/addon/macleod.cc
#include <nan.h>
#include <limits>
#include "macleod.h"

Nan::Persistent<v8::Function> MacLeod::constructor;

bool max(double a, double b) {
  if (b < a) return a;
  return b;
}

void MacLeod::Init(v8::Local<v8::Object> exports) {
  Nan::HandleScope scope;

  // Prepare constructor template
  v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(New);
  tpl->SetClassName(Nan::New("MacLeod").ToLocalChecked());
  tpl->InstanceTemplate()->SetInternalFieldCount(1);
  // Prototype
  Nan::SetPrototypeMethod(tpl, "getPitch", getPitch);

  constructor.Reset(tpl->GetFunction());
  exports->Set(Nan::New("MacLeod").ToLocalChecked(), tpl->GetFunction());
}

void MacLeod::New(const Nan::FunctionCallbackInfo<v8::Value>& info) {
  if (info.IsConstructCall()) {
    // Invoked as constructor: `new MyObject(...)`
    unsigned int bufferSize = info[0]->IsUndefined() ? DEFAULT_MACLEOD_BUFFER_SIZE : info[0]->NumberValue();
    double sampleRate = info[1]->IsUndefined() ? DEFAULT_MACLEOD_SAMPLE_RATE : info[1]->NumberValue();
    double cutoff = info[2]->IsUndefined() ? DEFAULT_MACLEOD_CUTOFF : info[2]->NumberValue();
    double freqCutoff = info[3]->IsUndefined() ? DEFAULT_MACLEOD_LOWER_PITCH_CUTOFF : info[3]->NumberValue();
    MacLeod* obj = new MacLeod(bufferSize, sampleRate, cutoff, freqCutoff);
    obj->Wrap(info.This());
    info.GetReturnValue().Set(info.This());
  } else {
    // Invoked as plain function `MyObject(...)`, turn into construct call.
    const int argc = 1;
    v8::Local<v8::Value> argv[argc] = { info[0] };
    v8::Local<v8::Function> cons = Nan::New<v8::Function>(constructor);
    info.GetReturnValue().Set(cons->NewInstance(argc, argv));
  }
}

void MacLeod::init(unsigned int bufferSize, double sampleRate, double cutoff, double freqCutoff) {
  this->nsdf = new double[bufferSize];
  this->bufferSize = bufferSize;
  this->sampleRate = sampleRate;
  this->cutoff = cutoff;
  this->lowerPitchCutoff = freqCutoff;
}

MacLeod::MacLeod() {
  init(DEFAULT_MACLEOD_BUFFER_SIZE, DEFAULT_MACLEOD_SAMPLE_RATE, DEFAULT_MACLEOD_CUTOFF, DEFAULT_MACLEOD_LOWER_PITCH_CUTOFF);
};
MacLeod::MacLeod(unsigned int bufferSize, double sampleRate, double cutoff, double freqCutoff) {
  init(bufferSize, sampleRate, cutoff, freqCutoff);
};
MacLeod::~MacLeod() {
  delete[] nsdf;
};

void MacLeod::getPitch(const Nan::FunctionCallbackInfo<v8::Value>& info) {
  MacLeod* obj = ObjectWrap::Unwrap<MacLeod>(info.Holder());
  assert(info[0]->IsFloat64Array());
  v8::Local<v8::Float64Array> input = info[0].As<v8::Float64Array>();
  assert(input->Length() <= obj->bufferSize);
  Nan::TypedArrayContents<double> inputData(input);
  info.GetReturnValue().Set(Nan::New(obj->calculatePitch((*inputData), input->Length())));
}

double MacLeod::normalizedSquareDifference(double* data, size_t dataSize) {
  for (size_t tau = 0; tau < dataSize; tau++) {
    double acf = 0;
    double divisorM = 0;
    for (size_t i = 0; i < dataSize - tau; i++) {
      acf += data[i] * data[i+tau];
      divisorM += data[i] * data[i] + data[i + tau] * data[i + tau];
    }
    nsdf[tau] = 2 * acf / divisorM;
  }
}

void MacLeod::parabolicInterpolation(unsigned int tau) {
  double nsdfa = nsdf[tau - 1];
  double nsdfb = nsdf[tau];
  double nsdfc = nsdf[tau + 1];
  double bValue = tau;
  double bottom = nsdfc + nsdfa - 2 * nsdfb;
  if (bottom == 0) {
    turningPointX = bValue;
    turningPointY = nsdfb;
  } else {
    double delta = nsdfa - nsdfc;
    turningPointX = bValue + delta / (2 * bottom);
    turningPointY = nsdfb - delta * delta / (8 * bottom);
  }
}

void MacLeod::peakPicking() {
  unsigned int pos = 0;
  unsigned int curMaxPos = 0;

  // find the first negative zero crossing.
  while (pos < (bufferSize - 1) / 3 && nsdf[pos] > 0) {
    pos++;
  }

  // loop over all the values below zero.
  while (pos < bufferSize - 1 && nsdf[pos] <= 0) {
    pos++;
  }

  // can happen if output[0] is NAN
  if (pos == 0) {
    pos = 1;
  }

  while (pos < bufferSize - 1) {
    if (nsdf[pos] > nsdf[pos - 1] && nsdf[pos] >= nsdf[pos + 1]) {
      if (curMaxPos == 0) {
        // the first max (between zero crossings)
        curMaxPos = pos;
      } else if (nsdf[pos] > nsdf[curMaxPos]) {
        // a higher max (between the zero crossings)
        curMaxPos = pos;
      }
    }
    pos++;
    // a negative zero crossing
    if (pos < bufferSize - 1 && nsdf[pos] <= 0) {
      // if there was a maximum add it to the list of maxima
      if (curMaxPos > 0) {
        maxPositions.push_back(curMaxPos);
        curMaxPos = 0; // clear the maximum position, so we start
        // looking for a new ones
      }
      while (pos < bufferSize - 1 && nsdf[pos] <= 0) {
        pos++; // loop over all the values below zero
      }
    }
  }
  if (curMaxPos > 0) {
    maxPositions.push_back(curMaxPos);
  }
}

double MacLeod::calculatePitch (double* data, size_t dataSize) {
  double pitch;
  maxPositions.clear();
  periodEstimates.clear();
  ampEstimates.clear();

  // 1. Calculute the normalized square difference for each Tau value.
  normalizedSquareDifference(data, dataSize);
  // 2. Peak picking time: time to pick some peaks.
  peakPicking();

  double highestAmplitude = - std::numeric_limits<double>::infinity();

  unsigned int tau;
  for (unsigned int i = 0; i < maxPositions.size(); i++) {
    tau = maxPositions[i];
    // make sure every annotation has a probability attached
    highestAmplitude = max(highestAmplitude, nsdf[tau]);

    if (nsdf[tau] > MACLEOD_SMALL_CUTOFF) {
      // calculates turningPointX and Y
      parabolicInterpolation(tau);
      // store the turning points
      ampEstimates.push_back(turningPointY);
      periodEstimates.push_back(turningPointX);
      // remember the highest amplitude
      highestAmplitude = max(highestAmplitude, turningPointY);
    }
  }

  if (periodEstimates.size() > 0) {
    // use the overall maximum to calculate a cutoff.
    // The cutoff value is based on the highest value and a relative
    // threshold.
    double actualCutoff = cutoff * highestAmplitude;
    unsigned int periodIndex = 0;

    for (unsigned int i = 0; i < ampEstimates.size(); i++) {
      if (ampEstimates[i] >= actualCutoff) {
        periodIndex = i;
        break;
      }
    }

    double period = periodEstimates[periodIndex];
    double pitchEstimate = sampleRate / period;

    if (pitchEstimate > lowerPitchCutoff) {
      pitch = pitchEstimate;
    } else {
      pitch = -1;
    }

  } else {
    // no pitch detected.
    pitch = -1;
  }

  probability = highestAmplitude;
  return pitch;
}
